### DISCUSSION TITLE
AVRbrain, a new AVR128DB64 dev board I'm working on

### DISCUSSION THREAD

**Discussion by jblang** [SpenceKonde/DxCore#515] [Show and tell]
I'm working on a new AVR128DB64 dev board that I plan to have manufactured soon ([schematic](https://github.com/SpenceKonde/DxCore/files/14060563/avrbrain.pdf)). I had to pre-order the AVR at JLCPCB and they are scheduled to arrive around Feb 6.  I'm planning to have the boards fabricated and partially assembled there (surface mount parts only).  This is my first surface mount board, and I welcome any feedback, especially about the surface mount stuff, but also generally.

![image](https://github.com/SpenceKonde/DxCore/assets/947942/46ff93a7-48c2-441b-841d-ff6c8e285a2b)

The board has the following features:

- Compatible with DxCore. I have tested the AVR128DB28 on a breadboard with the same [switchable programming circuit](https://github.com/SpenceKonde/DxCore/discussions/514#discussioncomment-8259341) I intend to use on my real board, and it works.
- Compatible with PlatformIO.  I will personally be using this to develop my own code because I hate the Arduino IDE. If I want to bang something together quickly using existing Arduino libraries, I will probably use DxCore with it. I have developed a large AVR code base* using pure avr-libc so I will probably be doing most of my development in that.  Here's the `platformio.ini` that I used to build a simple non-arduino led blinking example for the AVR128DB64:
  ```
  [env:AVR128DB28]
  platform = atmelmegaavr
  platform_packages = platformio/toolchain-atmelavr@^3.70300.220127
  board = AVR128DB28
  upload_speed = 115200
  upload_flags = 
	  --tool
	  uart
	  --device
	  avr128db28
	  --uart
	  $UPLOAD_PORT
	  --clk
	  $UPLOAD_SPEED
  upload_command = pymcuprog write --erase $UPLOAD_FLAGS --filename $SOURCE
  ```
  The above example can be made to work with DxCore just by adding the line `framework = arduino` to this file.
- Has an on board CH340C USB-to-serial converter with a micro-USB connector, and a switch to select between USB communication or programming via SerialUPDI.  
- The Optiboot_Dx bootloader can be used in USB comms mode, and there is an auto-reset circuit connected to the DTR line (possible to disable with a jumper).
- Has an on board 3.3V regulator for MVIO. The board is configured to have 3.3V on PORTC and 5V on the other ports, but this can be swapped, or the board can be made to use a single voltage by cutting some traces and adding solder bridges on the appropriate jumper pads.
- Has onboard SD card holder connected to 3.3V SPI Bus, and an I2C RTC (PCF8563T) with battery backup via CR 1612 coin cell.
- Has an Olimex [UEXT port](https://www.olimex.com/Products/Modules/) that exports 3.3V UART, I2C, and SPI buses.  Many peripheral modules are available from Olimex.
- Has a Sparkfun [Qwiic I2C port](https://www.sparkfun.com/qwiic) to work with the Qwiic and Stemma QT ecosystems.
- Has headers breaking out 6 8-bit ports (ports A-E and G). Pin sockets can be used for stackable shield-style add-on boards, and pin headers can be used to connect to breadboards or elsewhere using IDC cables. Every signal has its own ground return opposite, so it should be possible to maintain pretty good signal integrity over IDC.
- Port F is used for the USB UART, SD card chip select (level shifted), and RTC interrupt and 32 kHz clock signals, and of course UPDI and Reset. All of the RTC signals (including I2C) can be disconnected by cutting the traces in the jumper pads, and this can be undone by adding a solder bridge.
- UPDI header if you want to use a PicKit for debugging.
- Same size as a Raspberry Pi Model B. The right-side mounting holes are on the corners though instead of toward the middle of the board, and there is no common placement of GPIO headers since they're not compatible anyway.
- The board is two layer, but I've tried hard to maintain a good ground plane.

\* The large native avr-libc codebase I speak of is [z80ctrl](https://github.com/jblang/z80ctrl), and I am simultaneously designing a more specialized board for it using the same parts as this dev board.  Instead of the GPIO being broken out straightforwardly to headers, it will interface with the [RCBus](https://smallcomputercentral.com/rcbus/), a standard primarily used by Z80-based retrocomputers. I will have both manufactured at the same time.  I'll post a follow-up reply with more details about this board.

Here are some screenshots of the back silkscreen and the copper layers....

![image-1](https://github.com/SpenceKonde/DxCore/assets/947942/010c5c9f-a0a1-417a-acf8-31428bda5aad)
![image](https://github.com/SpenceKonde/DxCore/assets/947942/e252a326-5414-4e5e-94e9-3a9e40bee1aa)
![image-1](https://github.com/SpenceKonde/DxCore/assets/947942/3e6717ea-0892-4d8f-a915-f048783da62f)

**Comment by jblang** [comment]
Here is a [schematic](https://github.com/SpenceKonde/DxCore/files/14060686/z80ctrlng.pdf) and some 3D renders and screenshots of the copper layers for my new [z80ctrl](https://github.com/jblang/z80ctrl) board that I am designing simultaneously with the general purpose dev board using the same parts but a different pinout.  It provides a boot loader, bus-level debugger, and peripheral emulator for the Z80.  Instead of the GPIO being broken out straightforwardly to headers, it will interface with the [RCBus](https://smallcomputercentral.com/rcbus/), a standard used by Z80-based retrocomputers.

![image](https://github.com/SpenceKonde/DxCore/assets/947942/e2d43d0d-03db-4ca1-ac23-b726610794e6)
![image-1](https://github.com/SpenceKonde/DxCore/assets/947942/5028ccf1-672d-434c-8aee-cb7b20293f3b)
![image](https://github.com/SpenceKonde/DxCore/assets/947942/1343587d-7067-4f9e-b225-3e99352aebe7)
![image-1](https://github.com/SpenceKonde/DxCore/assets/947942/04e355c4-435e-408a-92ac-237c9657b5a5)

**Comment by jblang** [comment]
And finally, here are [schematics](https://github.com/SpenceKonde/DxCore/files/14060730/z80blinken.pdf), 3D renders, and copper layers for the companion z80blinken board, which holds the Z80 CPU and blinkenlights for all the control signals.  The z80ctrl plus this board and a suitable RCbus backplane are all it takes to have a full-functional Z80 computer that can run CP/M.

![image](https://github.com/SpenceKonde/DxCore/assets/947942/e07c51b1-07f3-4a4f-8030-615366b82202)
![image-1](https://github.com/SpenceKonde/DxCore/assets/947942/4210661c-31c5-4ebb-9ad4-0e7d449d5937)
![image](https://github.com/SpenceKonde/DxCore/assets/947942/b6e76a4e-6002-4b27-827b-034cb0f75101)
![image-1](https://github.com/SpenceKonde/DxCore/assets/947942/7eccc73f-8267-4f6b-9e67-f48bc1ff7ae0)

**Comment by microPaul** [comment]
I'm always a bit reluctant to comment on other people's designs, but since you asked....

Looks like you have a very nice AVR board here.  I'm guessing that the Red Xs on the schematic are to tell the board house the components you do *not* want them to assemble on the board (and you will do it yourself later).  Is that correct or is there some other meaning for those Xs?

I use the CH340C on my projects as well and I've had no problem with them (so far).  But I think Spence may have had some issues with the C part, so if you intend to do a large run of boards at some point you may want to check with Spence to understand the problems he (may have) had with the C parts.

For UPDI programming from a serial port, my own preference is to have a 4.7K resistor in parallel with the Schottky diode, to provide some additional pullup current on the RX lead that is sourced from the TX output pin.

My own preference for the Reset net is to have an additional pullup resistor and parallel diode (cathode to VDD) to absorb the charge from the driving cap when the net driving the cap goes high.  One can argue that the extra diode is not needed since the ESD diodes in the MCU may kick in before the external diode does (but then again maybe the external diode kicks in before the ESD diodes do).

For voltage regulators on my boards I'm moving away from the 1117 and to the AP7370 (Diodes Inc).  I don't like the need for the big capacitors (22uF) for the 1117 and because I'm trying to make all my boards capable of deep sleep with an idle current of less than 10 uA (5 uA is the target), including the regulator.  I like the AP7370 because it will tolerate up to 18V on the input and the idle current at no load is about 2 uA. and then scales up as the downstream devices draw more current from the regulator - so it looks like a great regulator for a sleeping device.   However, the 1117 will dissipate a lot more heat than the AP7370 can (the 1117 is physically larger and with a larger heatsink "tab").  So when the system regulator has a high voltage differential and substantive current draw, the 1117 can be a better choice than the AP7370.

I'd "feel" better if there were one or two 1uF caps physically close to the power pins on the MCU.   Not sure it's needed, but it would be easy to add to a future revision.

I'd also "feel" better if there was more stitching between the overlapping top and bottom ground plane areas.  Not sure it's needed, but it would be easy to add to a future revision.

I wish you the best with your new board.

Paul

**Reply by jblang** [reply]
Thanks for your feedback, I really do appreciate it.

>  Is that correct or is there some other meaning for those Xs?

Yes, that's because I had checked do not populate in KiCad.  I only want to pay JLCPCB for surface mount assembly.

>  But I think Spence may have had some issues with the C part, so if you intend to do a large run of boards at some point you may want to check with Spence to understand the problems he (may have) had with the C parts.

For now, I will stick with the C since I already ordered it, but will keep that in mind if I have problems.

> For voltage regulators on my boards I'm moving away from the 1117 and to the AP7370 (Diodes Inc).

I picked the 1117 because it is on JLCPCB's preferred parts list which means they won't charge you to load the pick and place. I am not planning to use the board in battery powered applications but if I do in the future, I may do another rev with the AP7370.

I will make the other tweaks you suggested regarding the caps and and resistors before I have this made.  On the vias, I have a return via next to all the signal vias so I think this should probably be sufficient. I may add some stitching though since it doesn't hurt anything.

**Comment by aemmenet** [comment]
My work in progress: https://github.com/aemmenet/polylab

**Reply by jblang** [reply]
Nice! What are the right angle connectors?

**Reply by aemmenet** [reply]
- 2 x upright 4-pin header with Grove Digital/Analog pinout

I added this connector description to the github README.md ... Thanks for asking ...

**Comment by TheRealDevonCalvin** [comment]
> I'm working on a new AVR128DB64 dev board that I plan to have manufactured soon ([schematic](https://github.com/SpenceKonde/DxCore/files/14060563/avrbrain.pdf)). I had to pre-order the AVR at JLCPCB and they are scheduled to arrive around Feb 6. I'm planning to have the boards fabricated and partially assembled there (surface mount parts only). This is my first surface mount board, and I welcome any feedback, especially about the surface mount stuff, but also generally.
> 
> ![image](https://private-user-images.githubusercontent.com/947942/299883334-46ff93a7-48c2-441b-841d-ff6c8e285a2b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU1MzgyNzksIm5iZiI6MTc0NTUzNzk3OSwicGF0aCI6Ii85NDc5NDIvMjk5ODgzMzM0LTQ2ZmY5M2E3LTQ4YzItNDQxYi04NDFkLWZmNmM4ZTI4NWEyYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyNFQyMzM5MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZTY4Nzg2ZWU0NjNhYjQ1ZTE1ZDhmMmQ0ZjM5NTI3MzBkMGNkNjY2OTI1YjIzMTQ0ZDhjNTM4MmQ1ZDBlMzhhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.G4ZjEZLZ8tjGinA8njojXq6hLWIhx2etXbK6FfqSzEc)
> 
> The board has the following features:
> 
>     * Compatible with DxCore. I have tested the AVR128DB28 on a breadboard with the same [switchable programming circuit](https://github.com/SpenceKonde/DxCore/discussions/514#discussioncomment-8259341) I intend to use on my real board, and it works.
> 
>     * Compatible with PlatformIO.  I will personally be using this to develop my own code because I hate the Arduino IDE. If I want to bang something together quickly using existing Arduino libraries, I will probably use DxCore with it. I have developed a large AVR code base* using pure avr-libc so I will probably be doing most of my development in that.  Here's the `platformio.ini` that I used to build a simple non-arduino led blinking example for the AVR128DB64:
>       ```
>       [env:AVR128DB28]
>       platform = atmelmegaavr
>       platform_packages = platformio/toolchain-atmelavr@^3.70300.220127
>       board = AVR128DB28
>       upload_speed = 115200
>       upload_flags = 
>           --tool
>           uart
>           --device
>           avr128db28
>           --uart
>           $UPLOAD_PORT
>           --clk
>           $UPLOAD_SPEED
>       upload_command = pymcuprog write --erase $UPLOAD_FLAGS --filename $SOURCE
>       ```
>       
>       
>           
>             
>           
>       
>             
>           
>       
>           
>         
>       The above example can be made to work with DxCore just by adding the line `framework = arduino` to this file.
> 
>     * Has an on board CH340C USB-to-serial converter with a micro-USB connector, and a switch to select between USB communication or programming via SerialUPDI.
> 
>     * The Optiboot_Dx bootloader can be used in USB comms mode, and there is an auto-reset circuit connected to the DTR line (possible to disable with a jumper).
> 
>     * Has an on board 3.3V regulator for MVIO. The board is configured to have 3.3V on PORTC and 5V on the other ports, but this can be swapped, or the board can be made to use a single voltage by cutting some traces and adding solder bridges on the appropriate jumper pads.
> 
>     * Has onboard SD card holder connected to 3.3V SPI Bus, and an I2C RTC (PCF8563T) with battery backup via CR 1612 coin cell.
> 
>     * Has an Olimex [UEXT port](https://www.olimex.com/Products/Modules/) that exports 3.3V UART, I2C, and SPI buses.  Many peripheral modules are available from Olimex.
> 
>     * Has a Sparkfun [Qwiic I2C port](https://www.sparkfun.com/qwiic) to work with the Qwiic and Stemma QT ecosystems.
> 
>     * Has headers breaking out 6 8-bit ports (ports A-E and G). Pin sockets can be used for stackable shield-style add-on boards, and pin headers can be used to connect to breadboards or elsewhere using IDC cables. Every signal has its own ground return opposite, so it should be possible to maintain pretty good signal integrity over IDC.
> 
>     * Port F is used for the USB UART, SD card chip select (level shifted), and RTC interrupt and 32 kHz clock signals, and of course UPDI and Reset. All of the RTC signals (including I2C) can be disconnected by cutting the traces in the jumper pads, and this can be undone by adding a solder bridge.
> 
>     * UPDI header if you want to use a PicKit for debugging.
> 
>     * Same size as a Raspberry Pi Model B. The right-side mounting holes are on the corners though instead of toward the middle of the board, and there is no common placement of GPIO headers since they're not compatible anyway.
> 
>     * The board is two layer, but I've tried hard to maintain a good ground plane.
> 
> 
> * The large native avr-libc codebase I speak of is [z80ctrl](https://github.com/jblang/z80ctrl), and I am simultaneously designing a more specialized board for it using the same parts as this dev board. Instead of the GPIO being broken out straightforwardly to headers, it will interface with the [RCBus](https://smallcomputercentral.com/rcbus/), a standard primarily used by Z80-based retrocomputers. I will have both manufactured at the same time. I'll post a follow-up reply with more details about this board.
> 
> Here are some screenshots of the back silkscreen and the copper layers....
> 
> ![image-1](https://private-user-images.githubusercontent.com/947942/299883358-010c5c9f-a0a1-417a-acf8-31428bda5aad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU1MzgyNzksIm5iZiI6MTc0NTUzNzk3OSwicGF0aCI6Ii85NDc5NDIvMjk5ODgzMzU4LTAxMGM1YzlmLWEwYTEtNDE3YS1hY2Y4LTMxNDI4YmRhNWFhZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyNFQyMzM5MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00OWNmNWI5YTM1ZDk0Y2U0OWRkMTk2ZmUxZTRkNzMwOTlhMzhjMjkzYjZmMjVjZDA3M2ZhN2YyNWZhZDY2ZGUxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.tKQqhgJg3FBjaE7lT4gCDCQDdYU3RJOytKxl-SxNegM) ![image](https://private-user-images.githubusercontent.com/947942/299883406-e252a326-5414-4e5e-94e9-3a9e40bee1aa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU1MzgyNzksIm5iZiI6MTc0NTUzNzk3OSwicGF0aCI6Ii85NDc5NDIvMjk5ODgzNDA2LWUyNTJhMzI2LTU0MTQtNGU1ZS05NGU5LTNhOWU0MGJlZTFhYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyNFQyMzM5MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYTg0OTgyZjBmZmMxZGJkNDM4ODcyMDgyMTYxYjE3YmI4M2FjYjcwMTU1M2MyMDUwODA2MDIxNDhkYzIzMjc2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.D1cZXcVWQ0yvBbiHsYlerqoWrjg22ZSwX9YtAFdJ-7c) ![image-1](https://private-user-images.githubusercontent.com/947942/299883421-3e6717ea-0892-4d8f-a915-f048783da62f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU1MzgyNzksIm5iZiI6MTc0NTUzNzk3OSwicGF0aCI6Ii85NDc5NDIvMjk5ODgzNDIxLTNlNjcxN2VhLTA4OTItNGQ4Zi1hOTE1LWYwNDg3ODNkYTYyZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyNFQyMzM5MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iNTQzODRkYTFhNzlhZjVjNmU3Yzk4OWM5OTRjN2MyZmE3NWJhMTA2YmNlN2ZiYzI1NjY1ZjgxNzhlYWZlN2I2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.j1JbAqfPcno-fvYjB6umoePSj_4vmL37TAOA7Dc2b1A)

Hey, just wondering how the PCB turned out? I'm looking for something exactly like this right now.

**Comment by jblang** [comment]
I have had it made, but I ran into some issues with the USB chip. It doesn't get recognized as a valid device when I connect it to my PC. I need to do some debugging. I have desoldered the USB chip from one board and mounted it on a separate breakout board so I can test it and the AVR separately, but I haven't gotten around to doing that yet. Maybe now that I know someone else is interested, that will motivate me to do that soon.

**Reply by TheRealDevonCalvin** [reply]
Let me know if you do, I'd love to see if you can get it working.

**Reply by jblang** [reply]
I got kind of discouraged by difficult to diagnose nature of the issue but maybe I will feel inspired to poke at it again this weekend.

**Comment by microPaul** [comment]
Have  you been able to have "any" CH340C be recognized by your PC?  Could it be a driver problem rather than an IC problem?   You can see the topology of how I use the CH340C in most of my designs at the link below, in case it's useful to you.  Scroll down until you see the schematic image, then click on the schematic to open a PDF of the schematic.  The CH340C and its surround is in the upper right corner of the schematic.

https://ad7i.net/main/cw-zoomer/

**Reply by jblang** [reply]
Yeah, I have ch340 powered arduino clones that work fine on the same pc. I still haven't gotten back to debugging this yet unfortunately.
